### DISCUSSION TITLE
Layer Shell with fltk-rs

### DISCUSSION THREAD

**Discussion by Arthex1** [fltk-rs/fltk-rs#1424] [Q&A]
Hello there, I am trying to make a wallpaper daemon with fltk-rs, and wanted to use layer shell to set the window's layer to `Background`. Is this possible, and if yes, how? I understand fltk doesn't support layer-shell natively as it is too "specific" for the framework, but are there any workarounds? ( IE: Using `wayland-protocols` with `wayland_protocols_wlr` )

Thats all, thank you!

**Comment by MoAlyousef** [comment]
Hello

I’m not really familiar with the wayland protocol, so maybe someone else can chime in with more info. 

Normally you would be able to get the wayland surface from a window using the raw_handle method:
(This uses the wayland-client crate to do something useful with it)
```rust
use fltk::{prelude::*, *};
use wayland_client::Proxy;

fn main() {
    let a = app::App::default();
    let mut wind = window::Window::new(100, 100, 400, 300, "My Window");
    wind.end();
    wind.show();
    let ptr = wind.raw_handle() as _;
    let val = unsafe { wayland_client::backend::ObjectId::from_ptr(wayland_client::protocol::wl_surface::WlSurface::interface(), ptr).unwrap()} ;
    dbg!(val);
    a.run().unwrap();
}
```

**Reply by Arthex1** [reply]
Hello, thank you for your reply. However it seems your wayland client version is outdated, as ObjectId::from_ptr no longer exists.

**Reply by MoAlyousef** [reply]
I haven’t tried recently, but it was just an example that wind.raw_handle() would give a valid wayland surface. I’m not familiar with the layer shell protocol of certain wayland compositor libraries, but I assume you would use the surface you get from the window and get a layer surface which you would then use.

**Comment by Arthex1** [comment]
> I haven’t tried recently, but it was just an example that wind.raw_handle() would give a valid wayland surface. I’m not familiar with the layer shell protocol of certain wayland compositor libraries, but I assume you would use the surface you get from the window and get a layer surface which you would then use.

Unfortunately, the `from_ptr` function seems to be a part of the older APIs, and I am not sure how to get the surface object with the current version of smithay/wayland-client.

**Reply by MoAlyousef** [reply]
It’s still available:
https://github.com/Smithay/wayland-rs/blob/90a9ad1f8f1fdef72e96d3c48bdb76b53a7722ff/wayland-backend/src/sys/mod.rs#L51

But requires the client_system feature flag.

**Reply by Arthex1** [reply]
Oh, i was looking at the wayland_client crate instead. My bad.

```rust
    let handle = wind.raw_handle();

    let ptr = handle as _;
    let id = unsafe { ObjectId::from_ptr(wayland_client::protocol::wl_surface::WlSurface::interface(), ptr).unwrap() };
    let surface = WlSurface::from_id(&conn, id).unwrap();

```
So I have this, and it is raising a segmentation fault? Which i don't think should be happening.
![image](https://github.com/fltk-rs/fltk-rs/assets/87478573/c0afa184-f759-4573-b934-85d4cd2c4e4f)

**Reply by MoAlyousef** [reply]
Are you building with the use-wayland feature flag?

**Reply by Arthex1** [reply]
yes i am, though the "segmentation fault" seems to come from `app.run().unwrap()` 

![image](https://github.com/fltk-rs/fltk-rs/assets/87478573/e1af5239-7438-4248-913a-091d3b06040a)


Code:
```rust
  let app = fltk::app::App::default();
    let mut wind = Window::new(100, 100, 400, 300, "Hello from rust");
    wind.end();
    wind.show();

    let handle = wind.raw_handle();


    let ptr = handle as _;
    dbg!(ptr);
    let id = unsafe { ObjectId::from_ptr(wayland_client::protocol::wl_surface::WlSurface::interface(), ptr).unwrap() };
    let surface = WlSurface::from_id(&conn, id).unwrap();
  ```

**Reply by MoAlyousef** [reply]
I can't reproduce on my system!
Can you post your Cargo.toml?

**Comment by Arthex1** [comment]
> I can't reproduce on my system! Can you post your Cargo.toml?

Here you go 

```toml
[dependencies]
env_logger = "0.10.0"
fltk = { version = "1.4.6", features = ["use-wayland"] }

```

Its just the basic cargo project with fltk as a dependency. Also, I am on Sway. Not sure if that is the problem, But it seemed worth mentioning. I will try in gnome in a bit.

Edit: tried In gnome, Same issue occurred.

**Reply by MoAlyousef** [reply]
I can't help you. You don't list your full Cargo.toml, nor your full code. I tried on 2 systems running wayland and I can't reproduce!

**Reply by Arthex1** [reply]
Apologies, i understand that is an error on my part. and I should provide more information.

```rust
use fltk::{app, prelude::*, window::Window};
use wayland_backend::client::ObjectId;
use wayland_client::{protocol::wl_surface::WlSurface, Connection, Proxy};

fn main() {
    let app = app::App::default();
    let mut window = Window::new(0, 0, 500, 500, "t");
    let connection = Connection::connect_to_env().unwrap();

    window.end();
    window.show();

    let handle = window.raw_handle();
    let ptr = handle as _;

    let id = unsafe { ObjectId::from_ptr(WlSurface::interface(), ptr).unwrap() };

    let surface = WlSurface::from_id(&connection, id).unwrap();
    app.run().unwrap();
}
```
```toml
[package]
name = "fltk-wallpaper"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
fltk = { git = "https://github.com/fltk-rs/fltk-rs.git", features = ["use-wayland"]}
smithay-client-toolkit = "0.17.0"
wayland-client = "0.30.2"
env_logger = "0.10.0"
wayland-backend = { version = "0.1.2", features = ["client_system"] }
```
```bash
xkbcommon: ERROR: couldn't find a Compose file for locale "en_IN" (mapped to "en_IN.ISO8859-1")
[1]    6252 segmentation fault (core dumped)  cargo run
```

I updated to the git version, incase that fixed anything. but it didn't seem to. I will try my best to be as informative as possible.

**Reply by MoAlyousef** [reply]
That is strange, I tried your code (and added a few dbg! macros):
```rust
use fltk::{app, prelude::*, window::Window};
use wayland_backend::client::ObjectId;
use wayland_client::{protocol::wl_surface::WlSurface, Connection, Proxy};

fn main() {
    let app = app::App::default();
    let mut window = Window::new(0, 0, 500, 500, "t");
    let connection = Connection::connect_to_env().unwrap();

    window.end();
    window.show();

    
    let handle = window.raw_handle();
    let ptr = handle as _;

    dbg!(&connection);
    let id = unsafe { ObjectId::from_ptr(WlSurface::interface(), ptr).unwrap() };
    dbg!(&id);
    let surface = WlSurface::from_id(&connection, id).unwrap();
    dbg!(&surface);
    app.run().unwrap();
}
```
It runs fine. This is the output:
```
[fltk/examples/temp.rs:17] &connection = Connection {
    backend: Backend {
        backend: InnerBackend {
            inner: Inner {
                state: Mutex {
                    data: ConnectionState {
                        display: 0x00005630dcb67780,
                        evq: 0x0000000000000000,
                        display_id: ObjectId(wl_display@1),
                        last_error: None,
                        known_proxies: {},
                    },
                    poisoned: false,
                    ..
                },
                dispatch_lock: Mutex {
                    data: Dispatcher,
                    poisoned: false,
                    ..
                },
            },
        },
    },
}
[fltk/examples/temp.rs:19] &id = ObjectId(wl_surface@15)
[fltk/examples/temp.rs:21] &surface = WlSurface {
    id: ObjectId(wl_surface@15),
    version: 4,
    data: None,
    backend: WeakBackend {
        inner: WeakInnerBackend {
            inner: (Weak),
        },
    },
}
```

**Reply by Arthex1** [reply]
```bash
xkbcommon: ERROR: couldn't find a Compose file for locale "en_IN" (mapped to "en_IN.ISO8859-1")
[src/main.rs:23] &connection = Connection {
    backend: Backend {
        backend: InnerBackend {
            inner: Inner {
                state: Mutex {
                    data: ConnectionState {
                        display: 0x000055d8bccf0630,
                        evq: 0x0000000000000000,
                        display_id: ObjectId(wl_display@1),
                        last_error: None,
                        known_proxies: {},
                    },
                    poisoned: false,
                    ..
                },
                dispatch_lock: Mutex {
                    data: Dispatcher,
                    poisoned: false,
                    ..
                },
            },
        },
    },
}
[src/main.rs:25] &id = ObjectId(wl_surface@17)
[src/main.rs:27] &surface = WlSurface {
    id: ObjectId(wl_surface@17),
    version: 4,
    data: None,
    backend: WeakBackend {
        inner: WeakInnerBackend {
            inner: (Weak),
        },
    },
}
[1]    11295 segmentation fault (core dumped)  cargo run
```
I have the same output aswell, except `app.run()` somehow segfaults on `use-wayland`, if I don't use use-wayland, it works fine...?

**Reply by MoAlyousef** [reply]
The use-wayland feature flag basically creates a hybrid app that can run on both X11 and wayland natively. Without it it will use the xwayland emulation layer on wayland systems.

**Reply by Arthex1** [reply]
Unfortunately, i dont think using xwayland will work for me. So I will try maybe in a VM if this works.

**Reply by MoAlyousef** [reply]
Does a simple fltk-rs app also segfault on app.run() on wayland:
```rust
use fltk::{app, prelude::*, window::Window};

fn main() {
    let app = app::App::default();
    let mut window = Window::new(0, 0, 500, 500, "t");

    window.end();
    window.show();

    app.run().unwrap();
}
```

**Reply by Arthex1** [reply]
yeah, it does.

**Reply by MoAlyousef** [reply]
Are you able to run through a debugger and get a stack trace?

**Reply by Arthex1** [reply]
Uh, i have never did such a thing before. I imagine by "debugger", you mean rust-gdb?

**Reply by MoAlyousef** [reply]
Or gdb itself. 
```
gdb ./target/debug/app
```
Then continually run by type `r` whenever the debugger stops. 
When it says there was a crash or segfault, type `bt` to get a backtrace.

**Reply by Arthex1** [reply]
Right, so i did it. the segmentation fault occurs. 

```rust
#0  xkb_compose_state_reset (state=0x0) at ../libxkbcommon/src/compose/state.c:135
#1  0x000055555558b519 in Fl::focus(Fl_Widget*) ()
#2  0x000055555558b838 in fl_fix_focus() ()
#3  0x000055555558bb71 in Fl::handle_(int, Fl_Window*) ()
#4  0x00005555555b80bf in wl_keyboard_enter(void*, wl_keyboard*, unsigned int, wl_surface*, wl_array*) ()
#5  0x00007ffff78294f6 in ffi_call_unix64 () at ../src/x86/unix64.S:104
#6  0x00007ffff7825f5e in ffi_call_int (cif=cif@entry=0x7fffffffd510, fn=<optimized out>, rvalue=<optimized out>, avalue=<optimized out>, closure=closure@entry=0x0)
    at ../src/x86/ffi64.c:673
#7  0x00007ffff7828b73 in ffi_call (cif=cif@entry=0x7fffffffd510, fn=<optimized out>, rvalue=rvalue@entry=0x0, avalue=avalue@entry=0x7fffffffd5e0) at ../src/x86/ffi64.c:710
#8  0x00007ffff7f8e645 in wl_closure_invoke (closure=closure@entry=0x55555574e6e0, target=<optimized out>,
    target@entry=0x555555717610, opcode=opcode@entry=1, data=<optimized out>, flags=1) at ../wayland-1.22.0/src/connection.c:1025
#9  0x00007ffff7f8ee73 in dispatch_event (display=display@entry=0x5555556fb3f0, queue=0x5555556fb4e0) at ../wayland-1.22.0/src/wayland-client.c:1631
#10 0x00007ffff7f8f13c in dispatch_queue (queue=0x5555556fb4e0, display=0x5555556fb3f0) at ../wayland-1.22.0/src/wayland-client.c:1777
#11 wl_display_dispatch_queue_pending (display=0x5555556fb3f0, queue=0x5555556fb4e0) at ../wayland-1.22.0/src/wayland-client.c:2019
#12 0x00005555555b7503 in wayland_socket_callback(int, wl_display*) ()
#13 0x0000555555636fe2 in Fl_Unix_Screen_Driver::poll_or_select_with_delay(double) ()
#14 0x0000555555637e1b in Fl_Unix_System_Driver::wait(double) ()
#15 0x000055555558b221 in Fl::run() ()
#16 0x0000555555582b26 in fltk::app::rt::run () at src/app/rt.rs:19
#17 0x0000555555580da1 in fltk::app::App::run () at src/app/mod.rs:59
#18 0x000055555557b9d2 in fltk_wallpaper::main () at src/main.rs:142
```
but unlike cargo, which immediately exits upon occurring a seg fault, and no window pops up. this doesn't happen with gdb, the window still shows after the seg fault, which I find quite weird.


![image](https://github.com/fltk-rs/fltk-rs/assets/87478573/ad07b166-7884-4402-b970-91646f8037e9)


Here is the output of `r` command
```rust
Enable debuginfod for this session? (y or [n]) y
Debuginfod has been enabled.
To make this setting permanent, add 'set debuginfod enabled on' to .gdbinit.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
xkbcommon: ERROR: couldn't find a Compose file for locale "en_IN" (mapped to "en_IN.ISO8859-1")
[New Thread 0x7ffff6a186c0 (LWP 4101)]
[Thread 0x7ffff6a186c0 (LWP 4101) exited]
[New Thread 0x7ffff6a186c0 (LWP 4107)]
[New Thread 0x7ffff60d56c0 (LWP 4108)]
[Thread 0x7ffff6a186c0 (LWP 4107) exited]
[Thread 0x7ffff60d56c0 (LWP 4108) exited]
[New Thread 0x7ffff60d56c0 (LWP 4112)]
[New Thread 0x7ffff6a186c0 (LWP 4114)]
[Thread 0x7ffff60d56c0 (LWP 4112) exited]
[Thread 0x7ffff6a186c0 (LWP 4114) exited]

Thread 1 "fltk-wallpaper" received signal SIGSEGV, Segmentation fault.
```

**Reply by MoAlyousef** [reply]
It seems to be occuring in the #0  xkb_compose_state_reset function.
I think the error:
```
xkbcommon: ERROR: couldn't find a Compose file for locale "en_IN" (mapped to "en_IN.ISO8859-1")
```
has to do something with it.
It's neither fltk-rs nor FLTK.

**Reply by Arthex1** [reply]
Well, isn't that error just supposed to be a warning? Also, what would cause the error if it's not fltk? Maybe fltk uses some wayland package I don't have installed?

**Reply by MoAlyousef** [reply]
It might be. But these warnings/errors can lead to such crashes/segfaults. Example the gdk_pixbuf segfault on library mismatch. 
This seems to be a related issue:
https://github.com/kovidgoyal/kitty/issues/439

**Reply by Arthex1** [reply]
Since i could not find a fix online, I asked ChatGPT out of desperate measure. and it suggested me to use `localectl` and reboot. Which I did, and no more segfaulting, so I imagine the error was indeed caused by xkbcommon.

**Reply by MoAlyousef** [reply]
If you have the complete command that you ran I would be grateful. It might help other users running into the same issue.

**Reply by Arthex1** [reply]
`sudo localectl set-locale LANG=en_IN.UTF-8`

**Reply by MoAlyousef** [reply]
Thank you
